detox/src/realms/DetoxPrimaryContext.js

```diff
@@ -227,11 +237,18 @@ class DetoxPrimaryContext extends DetoxContext {
       return;
     }
 
-    if (this[_deviceAllocator]) {
-      this[_deviceAllocator].emergencyCleanup();
-      this[_deviceAllocator] = null;
+    for (const key of Object.keys(this[_deviceAllocators])) {
+      const deviceAllocator = this[_deviceAllocators][key];
+      delete this[_deviceAllocators][key];
+      try {
+        deviceAllocator.emergencyCleanup();
+      } catch (err) {
+        this[symbols.logger].error({ cat: 'device', err }, `Failed to emergency cleanup device allocator for ${key}`);
+      }
     }
 
+    this[_cookieAllocators] = {};
+
```

Comment:
Test (unit) coverage is a must here.

Reply:
Both cleanup and emergency cleanup were thoroughly covered. Thanks for bringing attention to that.